mono/metadata/object.c

```diff
@@ -7580,88 +7580,24 @@ mono_message_init (MonoDomain *domain,
 {
 	MONO_REQ_GC_UNSAFE_MODE;
 
-	static MonoClass *object_array_klass;
-	static MonoClass *byte_array_klass;
-	static MonoClass *string_array_klass;
-	mono_error_init (error);
-	MonoMethodSignature *sig = mono_method_signature (method->method);
-	MonoString *name;
-	MonoArray *arr;
-	int i, j;
-	char **names;
-	guint8 arg_type;
-
-	if (!object_array_klass) {
-		MonoClass *klass;
+	static MonoMethod *init_message_method = NULL;
 
-		klass = mono_array_class_get (mono_defaults.byte_class, 1);
-		g_assert (klass);
-		byte_array_klass = klass;
-
-		klass = mono_array_class_get (mono_defaults.string_class, 1);
-		g_assert (klass);
-		string_array_klass = klass;
-
-		klass = mono_array_class_get (mono_defaults.object_class, 1);
-		g_assert (klass);
-
-		mono_atomic_store_release (&object_array_klass, klass);
+	if (!init_message_method) {
+		init_message_method = mono_class_get_method_from_name (mono_defaults.mono_method_message_class, "InitMessage", 2);
```

Comment:
System.dll is handled there https://github.com/xamarin/xamarin-macios/blob/master/tools/linker/MobileMarkStep.cs#L278

Some of that code could be moved down to mono/tools/linker. It's already layer'ed to split mobile, core (XI and XM) and products (when needed).

There are also unit tests that should be moved down to mono, so the linker could be tested (it's not today) inside mono (and re-used by the SDK).